File: c++/Queue.cpp
/*
Queue : 선입선출(FIFO) : enqueue(추가) / dequeue(제거) / front(제거없이 가져오기)
      : isEmpty(비어있는지 확인) / size(큐의 크기)
*/
#include<iostream>

class Node {
    public : 
        int data;
        Node* next;

        Node(int value) : data(value), next(nullptr) {}
};

// Queue class : First In First Out
class Queue{
    private:
        Node* front;
        Node* rear;
    
    public:
        Queue() : front(nullptr), rear(nullptr) {}

        void push(int value){
            Node* newNode = new Node(value);
            if (rear == nullptr){
                front = rear = newNode;
            }
            else{
                rear->next = newNode;
                rear = newNode;
            }
        }

        void pop(){
            if(isempty()){
                std::cout << "Queue is empty.";
                return;
            }
            // else
            Node* temp = front;
            front = front->next;
            delete temp;

            if (front == nullptr){
                rear = nullptr;
            }
        }

        int Front(){
            if (isempty()){
                std::cout << "Queue is empty.";
                return -1;
            }
            // else
            return front->data;
        }

        bool isempty(){
            return front == nullptr;
        }

        int size(){
            int cnt = 0;;
            Node* curr = front;
            while (curr != nullptr){
                cnt++;
                curr = curr->next;
            }
            return cnt;
        }

        void display(){
            Node* curr = front;
            while (curr != nullptr){
                std::cout << curr->data << " ";
                curr = curr->next;
            }
            std::cout << "\n";
            
        }
        ~Queue(){
            while (!isempty()){
                pop();
            }
        }
};

int main(){
    Queue Q;
    Q.push(0);
    Q.push(10);
    Q.push(20);
    Q.push(30);
    Q.push(40);
    Q.push(50);

    Q.display();
    std::cout << "front element : " << Q.Front() << "\n";
    std::cout << "Queue size : " << Q.size() << "\n";

    while (!Q.isempty()){
        std::cout << Q.Front() << "\n";
        Q.pop();
    }
}

File: Python/Queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.front = None
        self.rear = None
        self.size = 0

    def is_empty(self):
        return self.size == 0

    def enqueue(self, data):
        new_node = Node(data)
        if self.is_empty():
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node
        self.size += 1

    # pop_left()
    def dequeue(self):
        if self.is_empty():
            raise Exception("Queue is empty")
        data = self.front.data
        self.front = self.front.next
        self.size -= 1
        if self.is_empty():
            self.rear = None
        return data

    def peek(self):
        if self.is_empty():
            raise Exception("Queue is empty")
        return self.front.data

    def __len__(self):
        return self.size

    def __str__(self):
        if self.is_empty():
            return "Queue is empty"
        current = self.front
        items = []
        while current:
            items.append(str(current.data))
            current = current.next
        return " -> ".join(items)

# Example usage:
if __name__ == "__main__":
    queue = Queue()
    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)

    print("Queue:", queue)  # Output: Queue: 1 -> 2 -> 3

    print("Dequeue:", queue.dequeue())  # Output: Dequeue: 1
    print("Peek:", queue.peek())  # Output: Peek: 2
    print("Queue size:", len(queue))  # Output: Queue size: 2
